src/auxinfo/participant.rs

```diff
@@ -107,23 +107,17 @@ impl ProtocolParticipant for AuxInfoParticipant {
 
         match message.message_type() {
             MessageType::Auxinfo(AuxinfoMessageType::R1CommitHash) => {
-                let (broadcast_option, messages) = self.handle_broadcast(rng, message)?;
-
-                // This is kind of a bastardization of the outcome type, but this is basically a
-                // conversion from the broadcast outcome type (with a
-                // BroadcastOutput) to the aux-info outcome type.
-                let broadcast_outcome = ProcessOutcome::from(None, messages);
-
-                match broadcast_option {
-                    // If the round one broadcast worked, process the round one message.
-                    Some(bmsg) => {
-                        let round_one_outcome =
-                            self.handle_round_one_msg(rng, &bmsg, main_storage)?;
-                        broadcast_outcome.consolidate(vec![round_one_outcome])
-                    }
-                    // Otherwise, finish the broadcast.
-                    None => Ok(broadcast_outcome),
-                }
+                let broadcast_outcome = self.handle_broadcast(rng, message)?;
+
+                // Process the broadcast outcome:
+                // - If it included an output, use it as the agreed-upon round one message.
+                // - If it included messages, return them from this function
+                let (broadcast_output, messages) = broadcast_outcome.into_parts();
```

Comment:
This pattern repeats a couple times, and I wonder if it's worth isolating it as a method on `ProcessOutcome`, something like:
```rust
fn handle(self, f: F) -> Self {
    let (output, messages) = self.into_parts();
    let outcome = match output {
        Some(msg) => f(rng, &msg, ...)?,
        None => Self::Incomplete,
    };
    outcome.with_messages(messages)
}
```
where `f` will need to be given the appropriate type to match the `self.handle_round` calls.

Reply:
Hm, I'm having trouble passing a method on the `ProtocolParticipant` as a function parameter. The `handle` function compiles, but when I try to call it, I can't pass `self.handle_round_x` as a parameter -- it things I'm trying to pass a field
```rust
attempted to take value of method `handle_round_one_msg` on type `&mut auxinfo::participant::AuxInfoParticipant`
   --> src/auxinfo/participant.rs:112:54
    |
112 |                 broadcast_outcome.convert(self, self.handle_round_one_msg, rng, main_storage)
    |                                                      ^^^^^^^^^^^^^^^^^^^^ method, not a field
    |
help: use parentheses to call the method
    |
112 |                 broadcast_outcome.convert(self, self.handle_round_one_msg(_, _, _), rng, main_storage)
    |                                                                          +++++++++
```

I can't really tell from the documentation on closures and `Fn`/`FnMut` whether they're only supposed to be used for independent functions.

Reply:
Hmm, I've played with it for a bit and can't get it working either. I think you want to use `Self::handle_round_one_msg` but then the type signature for `handle` becomes complicated / I can't get it to work. Might not be worth the effort then, apologies for the side-tracking!

Reply:
Oh, that helps actually! I just pushed a draft, will come back tomorrow to finish docs.

Reply:
Taking a quick look now.